planner/cascades/transformation_rules.go

```diff
@@ -1341,6 +1345,105 @@ func (r *MergeAggregationProjection) OnTransform(old *memo.ExprIter) (newExprs [
+		return false
+	}
+	if len(agg.GroupByItems) != 0 {
+		return false
+	}
+	if len(agg.AggFuncs) != 1 {
+		return false
+	}
+	// If there is only one aggFunc, we don't need to guarantee that the child of it is a data
+	// source, or whether the sort can be eliminated. This transformation won't be worse than previous.
+	// Make sure that the aggFunc are Max or Min.
+	if agg.AggFuncs[0].Name != ast.AggFuncMax && agg.AggFuncs[0].Name != ast.AggFuncMin {
+		return false
+	}
+	return true
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `max/min->X` to `max/min->limit->sort->sel->X`.
+func (r *EliminateSingleMaxMin) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	agg := old.GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	childGroup := old.GetExpr().Children[0]
+	ctx := agg.SCtx()
+	f := agg.AggFuncs[0]
+
+	// If there's no column in f.GetArgs()[0], we still need limit and read data from real table because the result should be NULL if the input is empty.
+	if len(expression.ExtractColumns(f.Args[0])) > 0 {
+		// If it can be NULL, we need to filter NULL out first.
+		if !mysql.HasNotNullFlag(f.Args[0].GetType().Flag) {
+			sel := plannercore.LogicalSelection{}.Init(ctx, agg.SelectBlockOffset())
+			isNullFunc := expression.NewFunctionInternal(ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), f.Args[0])
+			notNullFunc := expression.NewFunctionInternal(ctx, ast.UnaryNot, types.NewFieldType(mysql.TypeTiny), isNullFunc)
+			sel.Conditions = []expression.Expression{notNullFunc}
+			selExpr := memo.NewGroupExpr(sel)
+			selExpr.SetChildren(childGroup)
+			selGroup := memo.NewGroupWithSchema(selExpr, childGroup.Prop.Schema)
+			childGroup = selGroup
+		}
+
+		// Add top(1) operators.
+		// For max function, the sort order should be desc.
+		desc := f.Name == ast.AggFuncMax
+		var byItems []*plannercore.ByItems
+		byItems = append(byItems, &plannercore.ByItems{
+			Expr: f.Args[0],
+			Desc: desc,
+		})
+		top1 := plannercore.LogicalTopN{
+			ByItems: byItems,
+			Count:   1,
+		}.Init(ctx, agg.SelectBlockOffset())
+		top1Expr := memo.NewGroupExpr(top1)
+		top1Expr.SetChildren(childGroup)
+		top1Group := memo.NewGroupWithSchema(top1Expr, childGroup.Prop.Schema)
+		childGroup = top1Group
+	} else {
+		li := plannercore.LogicalLimit{Count: 1}.Init(ctx, agg.SelectBlockOffset())
+		liExpr := memo.NewGroupExpr(li)
+		liExpr.SetChildren(childGroup)
+		liGroup := memo.NewGroupWithSchema(liExpr, childGroup.Prop.Schema)
+		childGroup = liGroup
+	}
+
+	newAgg := agg
+	newAggExpr := memo.NewGroupExpr(newAgg)
+	// If no data in the child, we need to return NULL instead of empty. This cannot be done by sort and limit themselves.
+	// Since now there would be at most one row returned, the remained agg operator is not expensive anymore.
+	newAggExpr.SetChildren(childGroup)
+	newAggExpr.AddAppliedRule(r)
+	return []*memo.GroupExpr{newAggExpr}, true, false, nil
```

Comment:
I think we'd better take more consideration on whether to erase the old plan piece. Consider this situation:

```sql
create table t(a bigint, b bigint) -- no index
select max(a) from t;
```

1. The old plan maybe is: TableScan -> Aggregate(max(a))
2. The new plan with TopN is: TableScan -> TopN(a, 1) -> Aggregate(max(a))

Since both TableScan is a full range table scan, it's hard to tell whether the new plan is better than the old one. Maybe we should keep the old one in this situation.